src/Microsoft.Extensions.FileProviders.Physical/CachingFileSystemProvider.cs

```diff
@@ -0,0 +1,54 @@
+﻿using System;
+using Microsoft.Extensions.Primitives;
+
+namespace Microsoft.Extensions.FileProviders.Physical
+{
+    public class CachingFileProvider: IFileProvider, IDisposable
+    {
+        private readonly IFileProvider _baseProvider;
+        private readonly ConcurrentLruCache<string, IFileInfo> _fileCache;
+        private readonly ConcurrentLruCache<string, IDirectoryContents> _directoryCache;
+        private Func<string, IFileInfo> _getFileInfo;
+        private Func<string, IDirectoryContents> _getDirectoryContents;
+        private IDisposable _changeCallbackRegistration;
+
+        public CachingFileProvider(IFileProvider baseProvider, int fileCapacity, int directoryCapacity, string watchFilter)
+        {
+            _baseProvider = baseProvider;
+            _fileCache = new ConcurrentLruCache<string, IFileInfo>(fileCapacity);
+            _directoryCache = new ConcurrentLruCache<string, IDirectoryContents>(directoryCapacity);
+            if (watchFilter != null)
+            {
+                _changeCallbackRegistration = _baseProvider.Watch(watchFilter).RegisterChangeCallback(Invalidate, null);
```

Comment:
This could potentially be polling in the background (which does GetFileInfo) so relying on this does not necessarily help. Additionally adding any of this to a static in time provider (either an embedded file provider or a physical file system that does not change) is probably going to be more overhead. Should we describe what scenarios we're trying to address here? Caching for the file system is too broad a concept and maybe we need something more fine grained?